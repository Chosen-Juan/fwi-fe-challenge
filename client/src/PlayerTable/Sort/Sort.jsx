import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { sortPlayers } from '../../appState/actions';
import './Sort.scss';

export class Sort extends Component {
  static propTypes = {
    playerKey: PropTypes.string.isRequired,
    sortPlayers: PropTypes.func.isRequired,
    sort: PropTypes.shape({
      key: PropTypes.string,
      direction: PropTypes.string,
    }).isRequired,
  };

  state = {
    direction: '',
  };

  changeDirection = cb => {
    const { direction } = this.state;
    if (direction === 'asc') {
      this.setState({ direction: 'desc' }, cb);
    } else {
      this.setState({ direction: 'asc' }, cb);
    }
  };

  sort = () => {
    const { playerKey, sortPlayers } = this.props;
    this.changeDirection(() => {
      const direction = this.state.direction ? this.state.direction : 'asc';
      sortPlayers(playerKey, direction);
    });
  };

  getSymbol = () => {
    const { direction } = this.state;
    if (!direction) {
      return '↕';
    } else if (direction === 'asc') {
      return '↓';
    } else {
      return '↑';
    }
  };

  componentWillReceiveProps({ sort, playerKey }) {
    if (sort.key !== playerKey) {
      this.setState({ direction: '' });
    }
  }

  render() {
    return (
      <button
        className="button is-primary is-small is-outlined"
        onClick={this.sort}
      >
        {this.getSymbol()}
      </button>
    );
  }
}

const mapStateToProps = ({ players: { sort } }) => ({
  sort,
});

const mapDispatchToProps = {
  sortPlayers,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Sort);
