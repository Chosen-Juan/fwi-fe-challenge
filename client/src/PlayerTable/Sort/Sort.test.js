import React from 'react';
import { shallow } from 'enzyme';

import { Sort } from './Sort';

describe('Sort', () => {
  const sortPlayersMock = jest.fn();
  const defaultProps = {
    sort: {
      key: 'name',
      direction: 'asc',
    },
    playerKey: 'name',
    sortPlayers: sortPlayersMock,
  };
  const sort = shallow(<Sort {...defaultProps} />);

  it('should match snapshot', () => {
    expect(sort).toMatchSnapshot();
  });

  it('should call sort players', () => {
    const button = sort.find('.button');
    button.simulate('click');
    expect(sortPlayersMock).toHaveBeenCalledTimes(1);
  });

  it('should switch direction back and forth', () => {
    const button = sort.find('.button');
    expect(sort.state('direction')).toEqual('asc');
    button.simulate('click');
    expect(sort.state('direction')).toEqual('desc');
    button.simulate('click');
    expect(sort.state('direction')).toEqual('asc');
  });
});
