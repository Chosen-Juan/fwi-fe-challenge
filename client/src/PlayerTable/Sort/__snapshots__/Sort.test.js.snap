// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort should match snapshot 1`] = `
<button
  className="button is-primary is-small is-outlined"
  onClick={[Function]}
>
  â†•
</button>
`;
